### PR TITLE
Add DisplayMessage to TestContext

### LINKED ISSUE #4237: Expose method(s) to send message/warning/error
## Summary

Expose method(s) to send message/warning/error.

## Background and Motivation

VSTest and the new platform offer capabilities to send messages to a client (e.g. Test Explorer) with Info/Warning/Error "levels". As part of the fixture methods when some processing is taking a long time, it would be nice to emit messages.

For examples, in Playwright, they are doing many "slow" downloads and install during `[AssemblyInitialize]` and having a basic level of progress report would be appreciated by users.

## Proposed Feature

Add either one method with an enum for "level" or multiple methods to send these messages. Given we are using VSTest bridge we can send them all with `IMessageLogger`.

I am not sure what's the best naming to use so that's clear for users what's the difference between `TestContext.WriteXXX` methods and these new APIs.

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #4237

### CODE REVIEW

File: src/TestFramework/TestFramework.Extensions/MessageLevel.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestTools.UnitTesting;
+
+public enum MessageLevel
```
Comment (Evangelink) [inline src/TestFramework/TestFramework.Extensions/MessageLevel.cs:9]: The enum is pretty self-explanatory but please add xml doc comment.

**Review by Evangelink — APPROVED** [review]

File: src/Adapter/MSTest.TestAdapter/Execution/TestExecutionManager.cs [hunk 2]
```diff
@@ -451,7 +462,10 @@ private void ExecuteTestsWithTestRunner(
             // Run single test passing test context properties to it.
             IDictionary<TestProperty, object?> tcmProperties = TcmTestPropertiesProvider.GetTcmProperties(currentTest);
             Dictionary<string, object?> testContextProperties = GetTestContextProperties(tcmProperties, sourceLevelParameters);
-            UnitTestResult[] unitTestResult = testRunner.RunSingleTest(unitTestElement.TestMethod, testContextProperties);
+
+            // testRunner is in a different app domain. We cannot pass the testExecutionRecorder directly.
```
Comment (Evangelink) [inline src/Adapter/MSTest.TestAdapter/Execution/TestExecutionManager.cs]: I would change the wording to `could be in a different AppDomain`.

File: src/TestFramework/TestFramework.Extensions/MessageLevel.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    /// <summary>
+    /// The message will be displayed in warning level.
+    /// </summary>
+    Warning,
+
+    /// <summary>
+    /// The message will be displayed in error level.
+    /// </summary>
+    Error,
+}
```
Comment (Evangelink) [inline src/TestFramework/TestFramework.Extensions/MessageLevel.cs]:
```suggestion
/// <summary>
/// Specifies the severity level of messages displayed using the <see cref="TestContext.DisplayMessage"/> API.
/// </summary>
public enum MessageLevel
{
    /// <summary>
    /// The message will be displayed as informational, typically used for general updates or non-critical messages.
    /// </summary>
    Informational,

    /// <summary>
    /// The message will be displayed as a warning, indicating a potential issue or something requiring attention.
    /// </summary>
    Warning,

    /// <summary>
    /// The message will be displayed as an error, representing a significant issue or failure.
    /// </summary>
    Error,
}
```

File: test/UnitTests/MSTestAdapter.PlatformServices.UnitTests/Services/TestContextImplementationTests.cs [hunk 4]
```diff
@@ -389,4 +391,24 @@ public void GetResultFilesShouldReturnListOfAddedResultFiles()
...
+
+        _testContextImplementation = new TestContextImplementation(_testMethod.Object, new ThreadSafeStringWriter(null, "test"), _properties, messageLoggerMock.Object);
+        _testContextImplementation.DisplayMessage(MessageLevel.Informational, "InfoMessage");
+        _testContextImplementation.DisplayMessage(MessageLevel.Warning, "WarningMessage");
+        _testContextImplementation.DisplayMessage(MessageLevel.Error, "ErrorMessage");
+
+        Assert.AreEqual(3, list.Count);
+        Assert.AreEqual((TestMessageLevel.Informational, "InfoMessage"), list[0]);
+        Assert.AreEqual((TestMessageLevel.Warning, "WarningMessage"), list[1]);
+        Assert.AreEqual((TestMessageLevel.Error, "ErrorMessage"), list[2]);
```
Comment (Evangelink) [inline test/UnitTests/MSTestAdapter.PlatformServices.UnitTests/Services/TestContextImplementationTests.cs]: This is working, a more moq oriented check would be:

```suggestion
        var messageLoggerMock = new Mock<IMessageLogger>(MockBehavior.Strict);

        messageLoggerMock
            .Setup(l => l.SendMessage(It.IsAny<TestMessageLevel>(), It.IsAny<string>()));

        _testContextImplementation = new TestContextImplementation(_testMethod.Object, new ThreadSafeStringWriter(null, "test"), _properties, messageLoggerMock.Object);
        _testContextImplementation.DisplayMessage(MessageLevel.Informational, "InfoMessage");
        _testContextImplementation.DisplayMessage(MessageLevel.Warning, "WarningMessage");
        _testContextImplementation.DisplayMessage(MessageLevel.Error, "ErrorMessage");

		messageLoggerMock.Verify(x => x.SendMessage(TestMessageLevel.Informational, "InfoMessage"), Times.Once);
        messageLoggerMock.Verify(x => x.SendMessage(TestMessageLevel.Warning, "WarningMessage"), Times.Once);
        messageLoggerMock.Verify(x => x.SendMessage(TestMessageLevel.Error, "ErrorMessage"), Times.Once);
```

**Review by Evangelink — APPROVED** [review]

File: src/Adapter/MSTest.TestAdapter/Execution/TestExecutionManager.cs (hunk 2 above)
Comment (Youssef1313) [reply src/Adapter/MSTest.TestAdapter/Execution/TestExecutionManager.cs]:
```suggestion
            // testRunner could be in a different AppDomain. We cannot pass the testExecutionRecorder directly.
```

File: src/TestFramework/TestFramework.Extensions/MessageLevel.cs (hunk 3 above)
Comment (Youssef1313) [reply src/TestFramework/TestFramework.Extensions/MessageLevel.cs]: IIRC, cref for TestContext doesn't work as this project doesn't have access to it. The dependency direction is the other way around

Comment (Youssef1313) [reply src/TestFramework/TestFramework.Extensions/MessageLevel.cs]: I must be confusing different things. TestContext is actually accessible from here. Will update to use cref

**Review by Evangelink — APPROVED** [review]

**Pull request merged**